Create the logic for the associations in the app (#251)
echo-swent/echo: docs/integration-association.md

# Association Integration
## Desciption
This file try to define how the associations will be handled in our application.
Firstly, let see some user stories linked to associations. Then, how we can implement those functionalities :
## User Stories
- As a user interested in philosophy, I want to be able to find related associations so that I don't have to make too much effort.
- As a user potentially interested in joining JDRPoly, I want to be able to get enough information on this association so that I know if it interests me or not.
- As a curious user, I want to be able to discover new associations that I never thought of so that I can start a new activity.
- As a user, I want to be able to follow an association so that I can easily see news and events related to it.

> - As a user in charge of communication in my association, I want to be able to share latest news so that people can see what's happening in my association.

Reviewer: I think he meant latest feed instead of latest news from what I can understand from the description which follows.

> - As a user in charge of communication in my association, I want to be able to share latest news so that people can see what's happening in my association.

Author: You're right, let's restraint the scope to just show the events created by the association, so it's easier and not necessitate new implementation only for news of an association

- As a user in charge of communication in my association, I want to be able to create events and certified them with my association so that other users can see that my event is from my association.
## Implementation
### Dataclass Association
We already have a dataclass called Association. Currently, it has the following attributes : an "associationId", a "name" and a "description". We can add to this :
- "relatedTags: Set<Tag>", the set of all tags related to the association

> - "moderators: Set<User>", the set of people that have the right to publish actuality, modify association description, add members and create events in the name of the association

Reviewer 2: there is actually a relation on the database (already present), specifying the committeeMembers of an association (n-to-n relation between `UserProfile`s and `Association`s). For now the property is already available on the `UserProfile` as the attribute `committeeMember`. This is needed for the Event screen to know for which Association the currently logged in User has the rights to create events for (also validated by the RLS Policy for inserts and updates on Events). I don't think having it on the Association makes sense since it would mean we'd need to fetch and add it to every Event displayed (since Events include the Association object in the `organizer` field. So far I can't see any need to have that information available since we don't plan to display it.

Author: I was not sure on how the comiteeMember was handled, thanks for the explanation. It makes sense to remove this attribute from Association object, I will correct it

> - TO DISCUSS : "members: Set<User>", list of all approved members of the association

Reviewer 2: I don't think this is something we need or want. Our app is about events and discovering them, not association membership.

> - "e-mail : String", a reference to the e-mail of the association if users want to contact them

Reviewer: It may be nice to have a "photos" section linked as well, such that users can find the relevant pictures inside the app itself and they do not need to quit the app to see photos.

Author: We will add the website of each association, so people can follow the link if they want more details about this kind of things ;)

- "website: String? = "" ", a reference to the website of the association for those who have one

> - "instagram: String? = "" ", a reference to the instagram of the association for those who have one

Reviewer 2: I'd make that a generic URL field (or a list of urls actually), rather than website and instagram specifically. Users might want to link to other things like other types of social media as well.

### AssociationsScreen
In the hamburger menu, we can add a route to go to an AssociationsScreen where we can find everything related to associations. 

> The first thing a user see is the recent actuality of each associations she/he follows, like a feed (For simplicity, instead of Actuality, we can firstly just show all events created by the association). 

Reviewer 2: ```suggestion
The first thing a user sees is the recent news of each associations she/he follows, like a feed (For simplicity, instead of news, we can firstly just show all events created by the association). 
```
Again, I don't think our app is really about "news" in general. we should probably talk about events instead.

> On top of this page, we have two buttons : one with "Followed Associations" which leads to a list of followed associations and one with "My Associations" which leads to a list of associations in which the user belong. 

Reviewer 2: This should clearly not be buttons, but one of the material design tab interface component. (But that's a technical detail of course).

The second part of this obviously would need to be removed if we don't pursue the idea of having association membership in our data/app as discussed above.

Reviewer 2: ```suggestion
On top of this page, we have two buttons : one with "Followed Associations" which leads to a list of followed associations and one with "My Associations" which leads to a list of associations the user is a member of. 
```

Author: Not sure to see what you reference as "the second part of this..." ?

> At the bottom right, there is a search button (like the one in HomeScreen) which allows the user to search an association. 

Reviewer 2: ```suggestion
At the bottom right, there is a search button (like the one in HomeScreen) which allows the user to search for associations. 
```

### FollowedAssociationsPage

> Already designed by Yoan (just needs little update), it consists of the list of all followed associations under the form of a list. Each item of the list have an unsubscribe button. If an item of the list is clicked, it leads the user to the specific AssociationPage.

Reviewer 2: ```suggestion
Already designed by Yoan (just needs a small update). It consists of the list of all followed associations under the form of a list. Each item of the list has an unsubscribe button. If an item of the list is clicked, it leads the user to the specific AssociationPage.
```

I'm actually not convinced that having the unsubscribe buttons this prominently is actually a thing we want. It makes it easy for users to click them accidentally, maybe without noticing. Mass unfollowing is probably a really rare thing for a user to do, so only allowing tu unfollow from the AssociationPage would be better. cc @unglazedstamp

> ### ComiteeAssociationsPage

Reviewer 2: ```suggestion
### CommitteeAssociationsPage
```

> Already designed by Yoan (just needs little update), it consists of the list of all associations in which the user belong. If an item of the list is clicked, it leads the user to the specific AssociationPage.

Reviewer 2: ```suggestion
Already designed by Yoan (just needs a small update), it consists of the list of all associations the user is a committee member of. If an item of the list is clicked, it leads the user to the specific AssociationPage.
```

### SearchMenuAssociations

> When the search button is clicked, it opens a bottom sheet with exactly the same functionnality as Discover, but without the filters (which doesn't make sense to search associations). A list with all associations is displayed in the background and updated according to the bottom sheet. If an item of the list is clicked, it leads the user to the specific AssociationPage.

Reviewer 2: ```suggestion
When the search button is clicked, it opens a bottom sheet with exactly the same functionnality as Discover, but without the filters (which don't make sense for searching associations). A list with all associations is displayed in the background and updated according to the bottom sheet. If an item of the list is clicked, it leads the user to the specific AssociationPage.
```

> When the search button is clicked, it opens a bottom sheet with exactly the same functionnality as Discover, but without the filters (which doesn't make sense to search associations). A list with all associations is displayed in the background and updated according to the bottom sheet. If an item of the list is clicked, it leads the user to the specific AssociationPage.

Reviewer 2: if there are no filters, is there a real need for the bottom sheet left ? if it's just a search bar with not much more, I'd rather opt for a simple search bar opening in the top app bar, as done by a lot of apps.

Author: The idea is to answer the user story of a student who wants to find associations linked to philosophy (for example). So we have the search bar and the discover tool. Check how it works in the latest version of the app for events filtering ;)

### AssociationPage

> On top, we have the choice of two pages : a "Description" default page and a "Actuality" page.

Reviewer 2: ```suggestion
On top, we have the choice of two pages : a "Description" default page and a "News" page.
```

#### DescriptionPage
This page consists of the title and description of the association. On top right, there is a button to follow/unfollow the association. If we scroll down, we can see a link to the association website, then a link to the instagram, and an e-mail to contact if the user is interested in joining the association (or for whatever question). If the user belongs to the moderators, she/he has an additional "Edit" button to modify the displayed information.
#### ActualityPage

> This page shows the history of all posts of the association (like a feed). If the user belongs to the moderators, she/he can post new actualities in the associtations feed for the followers.(For simplicity, instead of Actuality, we can firstly just show all events created by the association)

Reviewer: You mean you just want to implement it such that it only shows all the activities done by the association sorted by time for the association for now?

Author: Yes, and as we discussed in the meeting, we won't have news for association but only the events created by it
